jina/types/document/graph.py

```diff
@@ -0,0 +1,339 @@
+            with ImportExtensions(
+                required=False,
+                pkg_name="scipy",
+                help_text=f"GraphDocument needs scipy ",
+            ):
+                import scipy
+
+                JINA_GLOBAL.scipy_installed = True
+
+    def add_node(self, node: "Document"):
+        """
+        Add a a node to the graph
+
+        :param node: the node to be added to the graph
+        """
+        if node.id in self._node_id_to_offset:
+            default_logger.warning(f"Document {node.id} is already a node of the graph")
+            return
+
+        self._node_id_to_offset[node.id] = len(self.nodes)
+        self.nodes.append(node)
+
+    def remove_node(self, node: "Document"):
+        """
+        Remove a node from the graph along with the edges that may contain it
+
+        :param node: the node to be removed from the graph
+        """
+        from scipy.sparse import coo_matrix
+
+        if node.id not in self._node_id_to_offset:
+            default_logger.warning(
+                f"Trying to remove document {node.id} from the graph while is not a node of the graph"
+            )
+            return
+
+        offset = self._node_id_to_offset[node.id]
+
+        if self.num_edges > 0:
+            edges_to_remove = []
+            for edge_id, (row, col) in enumerate(
+                zip(self.adjacency.row, self.adjacency.col)
+            ):
+                if row.item() == offset or col.item() == offset:
+                    edge_features_keys = (
+                        f"{self.nodes[row.item()].id}-{self.nodes[col.item()]}"
+                    )
+                    edges_to_remove.append((edge_id, edge_features_keys))
+
+            for edge_id, edge_features_key in reversed(edges_to_remove):
+                self._remove_edge_id(edge_id, edge_features_key)
+
+            if self.num_edges > 0:
+                row = np.copy(self.adjacency.row)
+                col = np.copy(self.adjacency.col)
+                data = np.copy(self.adjacency.data)
+                for i in range(self.num_edges):
+                    if self.adjacency.row[i] > offset:
+                        row[i] = row[i] - 1
+                    if self.adjacency.col[i] > offset:
+                        col[i] = col[i] - 1
+                self.adjacency = coo_matrix((data, (row, col)))
+
+        del self.nodes[offset]
+        self._node_id_to_offset = {
+            node.id: offset for offset, node in enumerate(self.nodes)
+        }
+
+    def add_edge(
+        self, doc1: "Document", doc2: "Document", features: Optional[Dict] = None
+    ):
+        """
+        Add an edge to the graph connecting `doc1` with `doc2`
+
+        :param doc1: the starting node for this edge
+        :param doc2: the ending node for this edge
+        :param features: Optional features dictionary to be added to this new created edge
+        """
+        from scipy.sparse import coo_matrix
+
+        for doc in [doc1, doc2]:
```

Comment:
why for loop here? isn't doc1 and doc2 a pair of document?

Reply:
yes, we could do `self.add_node` twice

Reply:
But there is still the need to loop right?

Reply:
or one after the other

Reply:
okay